{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "46ff5e5b",
   "metadata": {},
   "source": [
    "<!-- 1. All the genes to include are in ./shortgenes2design_full.tsv. Specifically the short genes are in ./selected_shortgenes.tsv. \n",
    "2. ppDesigner results for the short genes are ./shortgenes_PPD/probes_DpnII-filtered_ov20.ppd\n",
    "3. For the non-short (or long) genes, we have two sources from which to select probes. We will use the overlapping source if there are not enough non-overlapping probes\n",
    "    1. ../221021_WholeGenome_noOverlapProbes/probes_DpnII-filtered.ppd\n",
    "    2. ../221010_WholeGenome_overlappingProbes/probes_DpnII-filtered.ppd\n",
    "4. The probes are going to be 3-on 3-off\n",
    "5. Amplification primers are going to be AP1V6 and AP2V6 (AP1V6 is a modified version of AP1V7) -->\n",
    "1. sppDesigner run is here: /home/kkalhor/Codes/ppDesignerRuns/230725_hg38_overlap20/probes_DpnII-filtered.ppd\n",
    "2. gene to barcode assignments are here: \"./gene_assignments/best_run.tsv\"\n",
    "\n",
    "We are aiming for 50 probes per gene. For every gene, we make two subsets of probes: the larger, l-subset and the smaller, s-subset. The probes in each subset are mutually non-overlapping, but they will overlap across subsets. If the l-subset has more than ```n_max``` probes, we randomly sample ```n_max``` probes without replacement. Otherwise, if l-subset and s-subset together have more than ```n_max``` probes, the select all of l-subset and then top it off with random sampling from s-subset. If l-subset and s-subset together have less than ```n_max``` probes, select all of them and top off with random sampling. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "67e9796c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>Sox9</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Dcn</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>Pdgfra</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>Thy1</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>Cdh5</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>914</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Fosl1</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>916</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>Fosl2</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>918</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Atf3</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>924</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Atf4</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>925</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Ets1</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>311 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  0  1  2  3  4  5  6  7 Gene name Source\n",
       "0        0  1  2  0  0  0  0  0  3      Sox9  Mouse\n",
       "1        1  1  0  2  0  3  0  0  0       Dcn  Mouse\n",
       "2        2  0  0  1  0  0  1  0  2    Pdgfra  Mouse\n",
       "3        3  2  2  0  0  0  0  3  0      Thy1  Mouse\n",
       "4        4  0  0  3  3  0  0  2  0      Cdh5  Mouse\n",
       "..     ... .. .. .. .. .. .. .. ..       ...    ...\n",
       "309    914  2  0  0  2  0  1  0  0     Fosl1  Mouse\n",
       "310    916  0  0  1  0  0  0  3  2     Fosl2  Mouse\n",
       "311    918  2  0  1  0  0  0  1  0      Atf3  Mouse\n",
       "312    924  0  0  2  0  0  2  0  1      Atf4  Mouse\n",
       "313    925  0  0  3  0  0  2  1  0      Ets1  Mouse\n",
       "\n",
       "[311 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os, re\n",
    "import numpy as np, pandas as pd\n",
    "import pickle\n",
    "from matplotlib import pyplot as plt\n",
    "rng = np.random.seed(seed=0)\n",
    "\n",
    "n_max = 10 # maximum number of probes per gene\n",
    "target_len = 152 # add A's to the 5' of the final probes to get them to this length\n",
    "assign_df = pd.read_csv(\"./barcodes/240508_3on-5off-3col_Hamming2_network_panel.tsv\", sep=\"\\t\", index_col=0).dropna()\n",
    "empty_df = assign_df.loc[(assign_df['Gene name'].str.startswith('Empty'))]\n",
    "assign_df = assign_df.loc[(~assign_df['Gene name'].str.startswith('Empty'))&(assign_df['Source']!='Human')]\n",
    "allgenes = assign_df['Gene name'].unique()\n",
    "display(assign_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a8d028e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'mgi_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'protein_id', 'ccdsid', 'ont']\n"
     ]
    }
   ],
   "source": [
    "# determine the probe is in CDS\n",
    "import gtfparse, os, pandas as pd\n",
    "annot_file = os.path.join('/home/xli/reference/mouse/mm39/gencode.vM35.annotation.gtf')\n",
    "annot_df = gtfparse.read_gtf(annot_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "df5f21fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "refseq_df = pd.DataFrame(data=annot_df, columns=annot_df.columns)\n",
    "refseq_df = refseq_df.loc[(refseq_df['gene_name'].isin(assign_df['Gene name']))&\n",
    "                            (refseq_df['feature']=='CDS')&\n",
    "                            (refseq_df['tag'].str.contains('Ensembl_canonical'))] # only design probes targeting canonical transcripts\n",
    "refseq_df = refseq_df.groupby(['gene_name','feature','strand']).agg({'start': 'min', 'end': 'max'}).reset_index()\n",
    "# add ncRNA gene: Morrbid\n",
    "refseq_df = pd.concat([refseq_df,pd.DataFrame({'gene_name': 'Morrbid', 'feature': 'CDS', 'strand': '-', 'start': 128020239, 'end': 128271286}, index=[0])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "d006e63a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(assign_df['Gene name']) - set(refseq_df['gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "e71ad9f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>probe_center</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7019</th>\n",
       "      <td>Il17a_chr1_1</td>\n",
       "      <td>chr1:020801129-20801212</td>\n",
       "      <td>0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>GCAAACATGAGTCCA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>GGGAGAGCTTCATCT</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.370986</td>\n",
       "      <td>W</td>\n",
       "      <td>GCAAACATGAGTCCAGGGAGAGCTTCATCT</td>\n",
       "      <td>1.2</td>\n",
       "      <td>1s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020801129</td>\n",
       "      <td>20801212</td>\n",
       "      <td>20801194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7020</th>\n",
       "      <td>Il17a_chr1_2</td>\n",
       "      <td>chr1:020802320-20802531</td>\n",
       "      <td>0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>ATCATCCCTCAAAGCT</td>\n",
       "      <td>51.0</td>\n",
       "      <td>CAGCGTGTCCAAACACT</td>\n",
       "      <td>58.0</td>\n",
       "      <td>2.416665</td>\n",
       "      <td>W</td>\n",
       "      <td>ATCATCCCTCAAAGCTCAGCGTGTCCAAACACT</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020802320</td>\n",
       "      <td>20802531</td>\n",
       "      <td>20802389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7021</th>\n",
       "      <td>Il17a_chr1_2</td>\n",
       "      <td>chr1:020802320-20802531</td>\n",
       "      <td>0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>AAACACTGAGGCCA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>AGGACTTCCTCCAGAATGT</td>\n",
       "      <td>58.0</td>\n",
       "      <td>2.437995</td>\n",
       "      <td>W</td>\n",
       "      <td>AAACACTGAGGCCAAGGACTTCCTCCAGAATGT</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020802320</td>\n",
       "      <td>20802531</td>\n",
       "      <td>20802413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7022</th>\n",
       "      <td>Il17a_chr1_2</td>\n",
       "      <td>chr1:020802320-20802531</td>\n",
       "      <td>0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>GACTTCCTCCAGAATGT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>GAAGGTCAACCTCAAA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.539507</td>\n",
       "      <td>W</td>\n",
       "      <td>GACTTCCTCCAGAATGTGAAGGTCAACCTCAAA</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020802320</td>\n",
       "      <td>20802531</td>\n",
       "      <td>20802432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7023</th>\n",
       "      <td>Il17a_chr1_2</td>\n",
       "      <td>chr1:020802320-20802531</td>\n",
       "      <td>0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>AAGGTCAACCTCAAAGT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>CTTTAACTCCCTTGGC</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.531687</td>\n",
       "      <td>W</td>\n",
       "      <td>AAGGTCAACCTCAAAGTCTTTAACTCCCTTGGC</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020802320</td>\n",
       "      <td>20802531</td>\n",
       "      <td>20802450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471967</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>1052.0</td>\n",
       "      <td>1052.0</td>\n",
       "      <td>TTGCTGTGGGACTCAA</td>\n",
       "      <td>55.0</td>\n",
       "      <td>CAGCATTTGTTTCTGAA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.611300</td>\n",
       "      <td>W</td>\n",
       "      <td>TTGCTGTGGGACTCAACAGCATTTGTTTCTGAA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>943876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471968</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>1278.0</td>\n",
       "      <td>1278.0</td>\n",
       "      <td>ACCACTGCCCAACT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>CTATGCTTCTTGATACATG</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.552558</td>\n",
       "      <td>W</td>\n",
       "      <td>accactgcccaactCTATGCTTCTTGATACATG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>944102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471969</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>3086.0</td>\n",
       "      <td>3086.0</td>\n",
       "      <td>TGGAGATGAATAAGTTGT</td>\n",
       "      <td>50.0</td>\n",
       "      <td>AAGGGTTGAGGCTAA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.485002</td>\n",
       "      <td>W</td>\n",
       "      <td>tggagatgaataagttgtaagggttgaggctaa</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>945910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471970</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>3233.0</td>\n",
       "      <td>3233.0</td>\n",
       "      <td>TCCATGAGATAAGGCA</td>\n",
       "      <td>51.0</td>\n",
       "      <td>ACTGTAAGGCATTTTCT</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.493049</td>\n",
       "      <td>W</td>\n",
       "      <td>tccatgagataaggcaactgtaaggcattttct</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>946057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471971</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>3338.0</td>\n",
       "      <td>3338.0</td>\n",
       "      <td>TCTATAAAAAGCAGGCTG</td>\n",
       "      <td>52.0</td>\n",
       "      <td>ATCAACCCAGGTGAG</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.528349</td>\n",
       "      <td>W</td>\n",
       "      <td>tctataaaaagcaggctgatcaacccaggtgag</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>946162</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20195 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0                        1  2       3       4  \\\n",
       "7019      Il17a_chr1_1  chr1:020801129-20801212  0    65.0    65.0   \n",
       "7020      Il17a_chr1_2  chr1:020802320-20802531  0    69.0    69.0   \n",
       "7021      Il17a_chr1_2  chr1:020802320-20802531  0    93.0    93.0   \n",
       "7022      Il17a_chr1_2  chr1:020802320-20802531  0   112.0   112.0   \n",
       "7023      Il17a_chr1_2  chr1:020802320-20802531  0   130.0   130.0   \n",
       "...                ...                      ... ..     ...     ...   \n",
       "1471967  Kdm5d_chrY_26    chrY:000942824-946316  0  1052.0  1052.0   \n",
       "1471968  Kdm5d_chrY_26    chrY:000942824-946316  0  1278.0  1278.0   \n",
       "1471969  Kdm5d_chrY_26    chrY:000942824-946316  0  3086.0  3086.0   \n",
       "1471970  Kdm5d_chrY_26    chrY:000942824-946316  0  3233.0  3233.0   \n",
       "1471971  Kdm5d_chrY_26    chrY:000942824-946316  0  3338.0  3338.0   \n",
       "\n",
       "                          5     6                    7     8         9 10  \\\n",
       "7019        GCAAACATGAGTCCA  50.0      GGGAGAGCTTCATCT  51.0  2.370986  W   \n",
       "7020       ATCATCCCTCAAAGCT  51.0    CAGCGTGTCCAAACACT  58.0  2.416665  W   \n",
       "7021         AAACACTGAGGCCA  50.0  AGGACTTCCTCCAGAATGT  58.0  2.437995  W   \n",
       "7022      GACTTCCTCCAGAATGT  53.0     GAAGGTCAACCTCAAA  50.0  2.539507  W   \n",
       "7023      AAGGTCAACCTCAAAGT  53.0     CTTTAACTCCCTTGGC  51.0  2.531687  W   \n",
       "...                     ...   ...                  ...   ...       ... ..   \n",
       "1471967    TTGCTGTGGGACTCAA  55.0    CAGCATTTGTTTCTGAA  50.0  2.611300  W   \n",
       "1471968      ACCACTGCCCAACT  53.0  CTATGCTTCTTGATACATG  51.0  2.552558  W   \n",
       "1471969  TGGAGATGAATAAGTTGT  50.0      AAGGGTTGAGGCTAA  50.0  2.485002  W   \n",
       "1471970    TCCATGAGATAAGGCA  51.0    ACTGTAAGGCATTTTCT  50.0  2.493049  W   \n",
       "1471971  TCTATAAAAAGCAGGCTG  52.0      ATCAACCCAGGTGAG  51.0  2.528349  W   \n",
       "\n",
       "                                        11   12   13 gene_name   chr  \\\n",
       "7019        GCAAACATGAGTCCAGGGAGAGCTTCATCT  1.2   1s     Il17a  chr1   \n",
       "7020     ATCATCCCTCAAAGCTCAGCGTGTCCAAACACT  0.5   2s     Il17a  chr1   \n",
       "7021     AAACACTGAGGCCAAGGACTTCCTCCAGAATGT  0.5   2s     Il17a  chr1   \n",
       "7022     GACTTCCTCCAGAATGTGAAGGTCAACCTCAAA  0.5   2s     Il17a  chr1   \n",
       "7023     AAGGTCAACCTCAAAGTCTTTAACTCCCTTGGC  0.5   2s     Il17a  chr1   \n",
       "...                                    ...  ...  ...       ...   ...   \n",
       "1471967  TTGCTGTGGGACTCAACAGCATTTGTTTCTGAA  0.0  14s     Kdm5d  chrY   \n",
       "1471968  accactgcccaactCTATGCTTCTTGATACATG  0.0  14s     Kdm5d  chrY   \n",
       "1471969  tggagatgaataagttgtaagggttgaggctaa  0.0  14s     Kdm5d  chrY   \n",
       "1471970  tccatgagataaggcaactgtaaggcattttct  0.0  14s     Kdm5d  chrY   \n",
       "1471971  tctataaaaagcaggctgatcaacccaggtgag  0.0  14s     Kdm5d  chrY   \n",
       "\n",
       "        exon_start  exon_end  probe_center  \n",
       "7019     020801129  20801212      20801194  \n",
       "7020     020802320  20802531      20802389  \n",
       "7021     020802320  20802531      20802413  \n",
       "7022     020802320  20802531      20802432  \n",
       "7023     020802320  20802531      20802450  \n",
       "...            ...       ...           ...  \n",
       "1471967  000942824    946316        943876  \n",
       "1471968  000942824    946316        944102  \n",
       "1471969  000942824    946316        945910  \n",
       "1471970  000942824    946316        946057  \n",
       "1471971  000942824    946316        946162  \n",
       "\n",
       "[20195 rows x 19 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def readPPD(file):\n",
    "    ppd = pd.read_csv(file, sep=\"\\t\")\n",
    "    ppd['gene_name'] = ppd['0'].str.split('_').str[0]\n",
    "    ppd[['chr', 'exon_start', 'exon_end']] = ppd['1'].str.extract(r\"(chr\\S+):(\\d+)-(\\d+)\")\n",
    "    ppd['probe_center'] = ppd['exon_start'].astype(int) + ppd['3'].astype(int)\n",
    "    return ppd\n",
    "all_ppd = readPPD(\"/home/xli/workspaces/dartfish/ProbeDesign/Codes/ppDesignerRuns/240724_mm39_splintR_overlap20/probes_DpnII-filtered.ppd\")\n",
    "all_ppd = all_ppd.query(\"gene_name in @allgenes\")\n",
    "all_ppd\n",
    "# # set(short_ppd['gene']).difference(shortgenes)\n",
    "\n",
    "# overlap_ppd = readPPD(\"../221010_WholeGenome_overlappingProbes/probes_DpnII-filtered.ppd\")\n",
    "# noovrlp_ppd = readPPD(\"../221021_WholeGenome_noOverlapProbes/probes_DpnII-filtered.ppd\")\n",
    "# overlap_ppd = overlap_ppd.query('gene_name in @longgenes')\n",
    "# noovrlp_ppd = noovrlp_ppd.query('gene_name in @longgenes')\n",
    "# # overlap_ppd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "bd0ebc4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_ppd = all_ppd.loc[all_ppd['gene_name'].isin(refseq_df['gene_name'])]\n",
    "\n",
    "# Annotate probes to CDS, 5'UTR or 3'UTR, considering the strand\n",
    "Annotation = []\n",
    "for i, row in all_ppd.iterrows():\n",
    "    gene = row['gene_name']\n",
    "    gene_df = refseq_df.loc[refseq_df['gene_name']==gene]\n",
    "    if len(gene_df)==0:\n",
    "        Annotation.append('NA')\n",
    "    else:\n",
    "        if gene_df['strand'].values[0]=='+':\n",
    "            if row['probe_center']<gene_df['start'].values[0]:\n",
    "                Annotation.append(\"5'UTR\")\n",
    "            elif row['probe_center']>gene_df['end'].values[0]:\n",
    "                Annotation.append(\"3'UTR\")\n",
    "            else:\n",
    "                Annotation.append(\"CDS\")\n",
    "        else:\n",
    "            if row['probe_center']<gene_df['start'].values[0]:\n",
    "                Annotation.append(\"3'UTR\")\n",
    "            elif row['probe_center']>gene_df['end'].values[0]:\n",
    "                Annotation.append(\"5'UTR\")\n",
    "            else:\n",
    "                Annotation.append(\"CDS\")\n",
    "all_ppd['Annotation'] = Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d5ba2f0b-913d-4cd0-a5ea-8045d04b2d2a",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Must have equal len keys and value when setting with an iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[40], line 10\u001b[0m\n\u001b[1;32m      8\u001b[0m left_lens \u001b[38;5;241m=\u001b[39m ppd_g[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m5\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39mlen()\u001b[38;5;241m.\u001b[39mvalues \u001b[38;5;66;03m# length of the 5' targets\u001b[39;00m\n\u001b[1;32m      9\u001b[0m right_lens \u001b[38;5;241m=\u001b[39m ppd_g[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m7\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39mlen()\u001b[38;5;241m.\u001b[39mvalues \u001b[38;5;66;03m# length of the 3' targets\u001b[39;00m\n\u001b[0;32m---> 10\u001b[0m \u001b[43mppd_g\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43miloc\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mppd_g\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mdist2prev\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m \u001b[38;5;241m=\u001b[39m centers[\u001b[38;5;241m1\u001b[39m:] \u001b[38;5;241m-\u001b[39m centers[\u001b[38;5;241m0\u001b[39m:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m     11\u001b[0m ppd_g[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbackreach\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;66;03m# minimum distance between two probe to get no overlap\u001b[39;00m\n\u001b[1;32m     12\u001b[0m ppd_g\u001b[38;5;241m.\u001b[39miloc[\u001b[38;5;241m1\u001b[39m:, ppd_g\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mget_loc(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbackreach\u001b[39m\u001b[38;5;124m'\u001b[39m)] \u001b[38;5;241m=\u001b[39m right_lens[\u001b[38;5;241m0\u001b[39m:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m] \u001b[38;5;241m+\u001b[39m left_lens[\u001b[38;5;241m1\u001b[39m:]\n",
      "File \u001b[0;32m~/.conda/envs/image/lib/python3.12/site-packages/pandas/core/indexing.py:912\u001b[0m, in \u001b[0;36m_LocationIndexer.__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m    909\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_has_valid_setitem_indexer(key)\n\u001b[1;32m    911\u001b[0m iloc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mname \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miloc\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39miloc\n\u001b[0;32m--> 912\u001b[0m \u001b[43miloc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_setitem_with_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda/envs/image/lib/python3.12/site-packages/pandas/core/indexing.py:1946\u001b[0m, in \u001b[0;36m_iLocIndexer._setitem_with_indexer\u001b[0;34m(self, indexer, value, name)\u001b[0m\n\u001b[1;32m   1943\u001b[0m \u001b[38;5;66;03m# align and set the values\u001b[39;00m\n\u001b[1;32m   1944\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m take_split_path:\n\u001b[1;32m   1945\u001b[0m     \u001b[38;5;66;03m# We have to operate column-wise\u001b[39;00m\n\u001b[0;32m-> 1946\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_setitem_with_indexer_split_path\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1947\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1948\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_setitem_single_block(indexer, value, name)\n",
      "File \u001b[0;32m~/.conda/envs/image/lib/python3.12/site-packages/pandas/core/indexing.py:2002\u001b[0m, in \u001b[0;36m_iLocIndexer._setitem_with_indexer_split_path\u001b[0;34m(self, indexer, value, name)\u001b[0m\n\u001b[1;32m   1997\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(value) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_integer(info_axis):\n\u001b[1;32m   1998\u001b[0m         \u001b[38;5;66;03m# This is a case like df.iloc[:3, [1]] = [0]\u001b[39;00m\n\u001b[1;32m   1999\u001b[0m         \u001b[38;5;66;03m#  where we treat as df.iloc[:3, 1] = 0\u001b[39;00m\n\u001b[1;32m   2000\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_setitem_with_indexer((pi, info_axis[\u001b[38;5;241m0\u001b[39m]), value[\u001b[38;5;241m0\u001b[39m])\n\u001b[0;32m-> 2002\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   2003\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMust have equal len keys and value \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   2004\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwhen setting with an iterable\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   2005\u001b[0m     )\n\u001b[1;32m   2007\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m lplane_indexer \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(value) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39mindex):\n\u001b[1;32m   2008\u001b[0m     \u001b[38;5;66;03m# We get here in one case via .loc with a all-False mask\u001b[39;00m\n\u001b[1;32m   2009\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: Must have equal len keys and value when setting with an iterable"
     ]
    }
   ],
   "source": [
    "\"\"\" Find the l-subset and s-subset for each gene\"\"\"\n",
    "temp = []\n",
    "for g in set(allgenes) - set(['Ccl27a','Cbx3','Bnip3']):\n",
    "    ppd_g = all_ppd.query(\"gene_name == @g\").copy()\n",
    "    ppd_g = ppd_g.sort_values('probe_center')\n",
    "    ppd_g['dist2prev'] = np.inf\n",
    "    centers = ppd_g['probe_center'].values\n",
    "    left_lens = ppd_g['5'].str.len().values # length of the 5' targets\n",
    "    right_lens = ppd_g['7'].str.len().values # length of the 3' targets\n",
    "    ppd_g.iloc[1:, ppd_g.columns.get_loc('dist2prev')] = centers[1:] - centers[0:-1]\n",
    "    ppd_g['backreach'] = 0 # minimum distance between two probe to get no overlap\n",
    "    ppd_g.iloc[1:, ppd_g.columns.get_loc('backreach')] = right_lens[0:-1] + left_lens[1:]\n",
    "    realsubset = ['l']\n",
    "    for i, (ind, (d, r)) in enumerate(ppd_g[['dist2prev', 'backreach']].iterrows()):\n",
    "        if i == 0:\n",
    "            continue\n",
    "        if r > d: # current probe overlaps with the previous probe\n",
    "            if realsubset[i-1] == 'l': # previous probe is in l-subset\n",
    "                realsubset.append('s')\n",
    "                continue\n",
    "        # if we're here it means that reach < distance, the previous probe was in s-subset\n",
    "        realsubset.append('l')\n",
    "    ppd_g['subset'] = realsubset\n",
    "    temp.append(ppd_g)    \n",
    "all_ppd = pd.concat(temp)\n",
    "all_ppd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74ad4ecb",
   "metadata": {},
   "source": [
    "### Preparing the decoding sequences\n",
    "5 probes have DpnII cut site in them. We replace them with new probes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "58262c12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>probe sequence</th>\n",
       "      <th>decoder sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dc0-488</th>\n",
       "      <td>TGTATCGCGCTCGATTGGCA</td>\n",
       "      <td>TGCCAATCGAGCGCGATACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc0-cy3</th>\n",
       "      <td>CGTATCGGTAGTCGCAACGC</td>\n",
       "      <td>GCGTTGCGACTACCGATACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc0-cy5</th>\n",
       "      <td>ACGCTACGGAGTACGCCACT</td>\n",
       "      <td>AGTGGCGTACTCCGTAGCGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc1-488</th>\n",
       "      <td>TCTTGCGTGCGATACGGAGT</td>\n",
       "      <td>ACTCCGTATCGCACGCAAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc1-cy3</th>\n",
       "      <td>AACGGTATTCGGTCGTCATC</td>\n",
       "      <td>GATGACGACCGAATACCGTT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               probe sequence      decoder sequence\n",
       "Name                                               \n",
       "dc0-488  TGTATCGCGCTCGATTGGCA  TGCCAATCGAGCGCGATACA\n",
       "dc0-cy3  CGTATCGGTAGTCGCAACGC  GCGTTGCGACTACCGATACG\n",
       "dc0-cy5  ACGCTACGGAGTACGCCACT  AGTGGCGTACTCCGTAGCGT\n",
       "dc1-488  TCTTGCGTGCGATACGGAGT  ACTCCGTATCGCACGCAAGA\n",
       "dc1-cy3  AACGGTATTCGGTCGTCATC  GATGACGACCGAATACCGTT"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>probe sequence</th>\n",
       "      <th>decoder sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dc3-cy3</th>\n",
       "      <td>CGTTTGATCGTTCGACCGAG</td>\n",
       "      <td>CTCGGTCGAACGATCAAACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc4-cy5</th>\n",
       "      <td>CAGGGATCGGTCGAGTACGC</td>\n",
       "      <td>GCGTACTCGACCGATCCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc6-488</th>\n",
       "      <td>CACGCTTACGATCCCGCTAT</td>\n",
       "      <td>ATAGCGGGATCGTAAGCGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc7-488</th>\n",
       "      <td>TGATCCTCCGCGAATCCATG</td>\n",
       "      <td>CATGGATTCGCGGAGGATCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc10-cy5</th>\n",
       "      <td>CCACAGATCGTGACTATCGG</td>\n",
       "      <td>CCGATAGTCACGATCTGTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc4-ATTO647N</th>\n",
       "      <td>CAGGGATCGGTCGAGTACGC</td>\n",
       "      <td>GCGTACTCGACCGATCCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc10-ATTO647N</th>\n",
       "      <td>CCACAGATCGTGACTATCGG</td>\n",
       "      <td>CCGATAGTCACGATCTGTGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     probe sequence      decoder sequence\n",
       "Name                                                     \n",
       "dc3-cy3        CGTTTGATCGTTCGACCGAG  CTCGGTCGAACGATCAAACG\n",
       "dc4-cy5        CAGGGATCGGTCGAGTACGC  GCGTACTCGACCGATCCCTG\n",
       "dc6-488        CACGCTTACGATCCCGCTAT  ATAGCGGGATCGTAAGCGTG\n",
       "dc7-488        TGATCCTCCGCGAATCCATG  CATGGATTCGCGGAGGATCA\n",
       "dc10-cy5       CCACAGATCGTGACTATCGG  CCGATAGTCACGATCTGTGG\n",
       "dc4-ATTO647N   CAGGGATCGGTCGAGTACGC  GCGTACTCGACCGATCCCTG\n",
       "dc10-ATTO647N  CCACAGATCGTGACTATCGG  CCGATAGTCACGATCTGTGG"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def getDcName(rnd, cl):\n",
    "    if cl == 0:\n",
    "        raise ValueError('Value 0 not accepted for a probe color')\n",
    "    fl_cl = {'1' : '488', '2' : 'cy3', '3' : 'cy5'}\n",
    "    return \"dc{0}-{1}\".format(rnd, fl_cl[str(cl)])\n",
    "\n",
    "def reverse_complement(dna):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "    return ''.join([complement[base] for base in dna[::-1]])\n",
    "\n",
    "dcProbes_File = './Decoding_Probes.xlsx'\n",
    "dpnii_site = 'GATC'\n",
    "dcProbes_df = pd.read_excel(dcProbes_File, sheet_name='all decoding probes', index_col=0)\n",
    "display(dcProbes_df.head())\n",
    "display(dcProbes_df.loc[dcProbes_df['probe sequence'].str.contains(dpnii_site)])\n",
    "\n",
    "dcProbes_df = dcProbes_df.loc[~dcProbes_df['probe sequence'].str.contains(dpnii_site)]\n",
    "dcProbes_df = dcProbes_df.set_index(dcProbes_df.index.str.split('_').str[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2146003f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(311, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12000003</td>\n",
       "      <td>Sox9</td>\n",
       "      <td>TGTATCGCGCTCGATTGGCA</td>\n",
       "      <td>AACGGTATTCGGTCGTCATC</td>\n",
       "      <td>GCTCAGCCGGACGAGTAGAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10203000</td>\n",
       "      <td>Dcn</td>\n",
       "      <td>TGTATCGCGCTCGATTGGCA</td>\n",
       "      <td>CTACTTCGTCGCGTCAGACC</td>\n",
       "      <td>ACTCTACCGGCAATCGCGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>00100102</td>\n",
       "      <td>Pdgfra</td>\n",
       "      <td>AGAACTTGCGCGGATACACG</td>\n",
       "      <td>GAGTGTCGCGCAACTTAGCG</td>\n",
       "      <td>GCCACATCGACTCGGTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22000030</td>\n",
       "      <td>Thy1</td>\n",
       "      <td>CGTATCGGTAGTCGCAACGC</td>\n",
       "      <td>AACGGTATTCGGTCGTCATC</td>\n",
       "      <td>CTCTCGTAGCGTGCGATGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>00330020</td>\n",
       "      <td>Cdh5</td>\n",
       "      <td>GACGAACGGTCGAGATTTAC</td>\n",
       "      <td>AACTGCGACCGTCGGCTTAC</td>\n",
       "      <td>TCGTAACCCGTGCGAAGTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>20020100</td>\n",
       "      <td>Fosl1</td>\n",
       "      <td>CGTATCGGTAGTCGCAACGC</td>\n",
       "      <td>TCGTACTTCGACGGCACTCA</td>\n",
       "      <td>GAGTGTCGCGCAACTTAGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>00100032</td>\n",
       "      <td>Fosl2</td>\n",
       "      <td>AGAACTTGCGCGGATACACG</td>\n",
       "      <td>CTCTCGTAGCGTGCGATGAG</td>\n",
       "      <td>GCCACATCGACTCGGTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>20100010</td>\n",
       "      <td>Atf3</td>\n",
       "      <td>CGTATCGGTAGTCGCAACGC</td>\n",
       "      <td>AGAACTTGCGCGGATACACG</td>\n",
       "      <td>CTTGCGGCGACAGTCGAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>00200201</td>\n",
       "      <td>Atf4</td>\n",
       "      <td>CTACTTCGTCGCGTCAGACC</td>\n",
       "      <td>ACGTCTGCGTACCGGCTTAG</td>\n",
       "      <td>TTAGGTCGCCTACCGACTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>00300210</td>\n",
       "      <td>Ets1</td>\n",
       "      <td>GACGAACGGTCGAGATTTAC</td>\n",
       "      <td>ACGTCTGCGTACCGGCTTAG</td>\n",
       "      <td>CTTGCGGCGACAGTCGAACA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>311 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      barcode Gene name                     0                     1  \\\n",
       "0    12000003      Sox9  TGTATCGCGCTCGATTGGCA  AACGGTATTCGGTCGTCATC   \n",
       "1    10203000       Dcn  TGTATCGCGCTCGATTGGCA  CTACTTCGTCGCGTCAGACC   \n",
       "2    00100102    Pdgfra  AGAACTTGCGCGGATACACG  GAGTGTCGCGCAACTTAGCG   \n",
       "3    22000030      Thy1  CGTATCGGTAGTCGCAACGC  AACGGTATTCGGTCGTCATC   \n",
       "4    00330020      Cdh5  GACGAACGGTCGAGATTTAC  AACTGCGACCGTCGGCTTAC   \n",
       "..        ...       ...                   ...                   ...   \n",
       "309  20020100     Fosl1  CGTATCGGTAGTCGCAACGC  TCGTACTTCGACGGCACTCA   \n",
       "310  00100032     Fosl2  AGAACTTGCGCGGATACACG  CTCTCGTAGCGTGCGATGAG   \n",
       "311  20100010      Atf3  CGTATCGGTAGTCGCAACGC  AGAACTTGCGCGGATACACG   \n",
       "312  00200201      Atf4  CTACTTCGTCGCGTCAGACC  ACGTCTGCGTACCGGCTTAG   \n",
       "313  00300210      Ets1  GACGAACGGTCGAGATTTAC  ACGTCTGCGTACCGGCTTAG   \n",
       "\n",
       "                        2  \n",
       "0    GCTCAGCCGGACGAGTAGAT  \n",
       "1    ACTCTACCGGCAATCGCGTC  \n",
       "2    GCCACATCGACTCGGTCTAT  \n",
       "3    CTCTCGTAGCGTGCGATGAG  \n",
       "4    TCGTAACCCGTGCGAAGTGC  \n",
       "..                    ...  \n",
       "309  GAGTGTCGCGCAACTTAGCG  \n",
       "310  GCCACATCGACTCGGTCTAT  \n",
       "311  CTTGCGGCGACAGTCGAACA  \n",
       "312  TTAGGTCGCCTACCGACTGC  \n",
       "313  CTTGCGGCGACAGTCGAACA  \n",
       "\n",
       "[311 rows x 5 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes_df = assign_df.copy() #assign_df.drop('gene', axis=1).copy()\n",
    "barcodes_seq = pd.DataFrame()\n",
    "for i in barcodes_df.index:\n",
    "    nonZeros_bc = barcodes_df.drop(['index','Gene name','Source'], axis=1).loc[i, barcodes_df.loc[i] != 0].to_dict()\n",
    "    nonZero_rnds = list(nonZeros_bc.keys())\n",
    "    nonZeros_seq = {rnd : dcProbes_df['probe sequence'].loc[getDcName(rnd, nonZeros_bc[rnd])]\n",
    "                    for rnd in nonZero_rnds}\n",
    "    bc_str = \"\".join(barcodes_df.drop(['index','Gene name','Source'], axis=1).loc[i].astype(str))\n",
    "    \n",
    "    thisBarcodes_seq = pd.DataFrame({'barcode' : bc_str, 'Gene name' : barcodes_df.loc[i, 'Gene name']},\n",
    "                                    index = [i]) # a df with one row\n",
    "    for j, rnd in enumerate(nonZeros_seq):\n",
    "        thisBarcodes_seq.loc[:, j] = nonZeros_seq[rnd] # adding dcProbes to the df\n",
    "    barcodes_seq = pd.concat([barcodes_seq, thisBarcodes_seq])\n",
    "    \n",
    "print(barcodes_seq.shape)\n",
    "barcodes_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "33a0ca41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>906</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>Junb</td>\n",
       "      <td>Mouse</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  0  1  2  3  4  5  6  7 Gene name Source\n",
       "306    906  0  1  0  3  0  0  2  0      Junb  Mouse"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes_df.loc[barcodes_df['Gene name']=='Junb']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb969886",
   "metadata": {},
   "source": [
    "## Synthesizing the barcode sequence\n",
    "We want to make sure to have DpnII free probes. So we'll have to have a combination of barcodes for each probe that doesn't give us GATC. Since the left and right arms, common insert and the decoding probes don't have native DpnII sites in them, potential problem could arise at junctions. \n",
    "\n",
    "For each barcode, we synthesize the probe with the following recipe:\n",
    "5' -> AP1 - left arm - barcode i1 - anchor probe - barcode i2 - barcode i3 - right arm - AP2 -> 3'  \n",
    "i1...i3 are a permutation of {0, 1, 2} that give us a DpnII free full probe (excluding the native DpnII cut site of AP2)\n",
    "\n",
    "In addition, we want all probes to be at least 160 bases long (max is 162). I'll add a number of As before the common insert\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1e86e5f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anchorProbe = 20, rcaPrimer = 18\n"
     ]
    }
   ],
   "source": [
    "# commonInsert = 'CTTCAGCTTCCCGATATCCGACGG' # 61C\n",
    "anchorProbe = 'CTTCAGCTTCCCGATATCCG' # 55.1C\n",
    "rcaPrimer = reverse_complement('GATATCGGGAAGCTGAAG') # 49.6C\n",
    "print('anchorProbe = {0}, rcaPrimer = {1}'.format(len(anchorProbe), len(rcaPrimer)))\n",
    "# AP1V6 = 'GTCATATCGGTCACTGTT'  # AP1 directly goes on the probes. Order pAP1V61U G*G*GTCATATCGGTCACTGTU for amplification\n",
    "# AP2V6 = 'GGGTAGTGTGTATCCTGATC' # AP2 will have to be reverse complemented on the oligos. Order AP2V6 /5Phos/CACGGGTAGTGTGTATCCTG for amplification.\n",
    "# AP1V72 = 'GCAAGATTCTCGTCGAGT'  # AP1 directly goes on the probes. Order AP1V6U A*A*GCAAGATTCTCGTCGAG/3deoxyU/  for amplification\n",
    "# AP2V7 = 'TGTAAGGCACATCTCGGATC' # AP2 will have to be reverse complemented on the oligos. Order AP2V6 /5Phos/TGTAAGGCACATCTCGGATC for amplification.\n",
    "\n",
    "AP1V8 = 'TAATCTAGCGCGACGTCT'  # AP1 directly goes on the probes. Order AP1V8U T*C*TAATCTAGCGCGACGTCU for amplification\n",
    "AP2V8 = 'CCACAAGAGGCGCTATGATC' # AP2 will have to be reverse complemented on the oligos. Order AP2V8 /5Phos/CCACAAGAGGCGCTATGATC for amplification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "1cf366c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No DpnII free barcode for gene Col10a1, index 850042\n",
      "No DpnII free barcode for gene Gadd45g, index 1083083\n",
      "No probes for Ccl27a.\n",
      "No probes for Ccl27b.\n",
      "No probes for Bnip3.\n",
      "No DpnII free barcode for gene Sox4, index 1070233\n",
      "No DpnII free barcode for gene Sox4, index 1070235\n",
      "No probes for H3c6.\n",
      "No DpnII free barcode for gene Suv39h2, index 111771\n",
      "No DpnII free barcode for gene Kdm5b, index 68039\n",
      "No DpnII free barcode for gene Kdm5b, index 68102\n",
      "No DpnII free barcode for gene Kdm5b, index 68168\n",
      "No DpnII free barcode for gene Kdm5b, index 68207\n",
      "No probes for Cbx3.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>13</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>probe_center</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>full_probe</th>\n",
       "      <th>RNA_target_probe</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7019</th>\n",
       "      <td>Il17a_chr1_1</td>\n",
       "      <td>chr1:020801129-20801212</td>\n",
       "      <td>0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>GCAAACATGAGTCCA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>GGGAGAGCTTCATCT</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.370986</td>\n",
       "      <td>...</td>\n",
       "      <td>1s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020801129</td>\n",
       "      <td>20801212</td>\n",
       "      <td>20801194</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTTGGACTCATGTTTGCAGAACTTGCGCGG...</td>\n",
       "      <td>AGATGAAGCTCTCCCTGGACTCATGTTTGC</td>\n",
       "      <td>00102300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7020</th>\n",
       "      <td>Il17a_chr1_2</td>\n",
       "      <td>chr1:020802320-20802531</td>\n",
       "      <td>0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>ATCATCCCTCAAAGCT</td>\n",
       "      <td>51.0</td>\n",
       "      <td>CAGCGTGTCCAAACACT</td>\n",
       "      <td>58.0</td>\n",
       "      <td>2.416665</td>\n",
       "      <td>...</td>\n",
       "      <td>2s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020802320</td>\n",
       "      <td>20802531</td>\n",
       "      <td>20802389</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTAGCTTTGAGGGATGATAGAACTTGCGCG...</td>\n",
       "      <td>AGTGTTTGGACACGCTGAGCTTTGAGGGATGAT</td>\n",
       "      <td>00102300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7021</th>\n",
       "      <td>Il17a_chr1_2</td>\n",
       "      <td>chr1:020802320-20802531</td>\n",
       "      <td>0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>AAACACTGAGGCCA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>AGGACTTCCTCCAGAATGT</td>\n",
       "      <td>58.0</td>\n",
       "      <td>2.437995</td>\n",
       "      <td>...</td>\n",
       "      <td>2s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020802320</td>\n",
       "      <td>20802531</td>\n",
       "      <td>20802413</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTTGGCCTCAGTGTTTAGAACTTGCGCGGA...</td>\n",
       "      <td>ACATTCTGGAGGAAGTCCTTGGCCTCAGTGTTT</td>\n",
       "      <td>00102300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7022</th>\n",
       "      <td>Il17a_chr1_2</td>\n",
       "      <td>chr1:020802320-20802531</td>\n",
       "      <td>0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>GACTTCCTCCAGAATGT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>GAAGGTCAACCTCAAA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.539507</td>\n",
       "      <td>...</td>\n",
       "      <td>2s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020802320</td>\n",
       "      <td>20802531</td>\n",
       "      <td>20802432</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTACATTCTGGAGGAAGTCAGAACTTGCGC...</td>\n",
       "      <td>TTTGAGGTTGACCTTCACATTCTGGAGGAAGTC</td>\n",
       "      <td>00102300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7023</th>\n",
       "      <td>Il17a_chr1_2</td>\n",
       "      <td>chr1:020802320-20802531</td>\n",
       "      <td>0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>AAGGTCAACCTCAAAGT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>CTTTAACTCCCTTGGC</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.531687</td>\n",
       "      <td>...</td>\n",
       "      <td>2s</td>\n",
       "      <td>Il17a</td>\n",
       "      <td>chr1</td>\n",
       "      <td>020802320</td>\n",
       "      <td>20802531</td>\n",
       "      <td>20802450</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTACTTTGAGGTTGACCTTAGAACTTGCGC...</td>\n",
       "      <td>GCCAAGGGAGTTAAAGACTTTGAGGTTGACCTT</td>\n",
       "      <td>00102300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471967</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>1052.0</td>\n",
       "      <td>1052.0</td>\n",
       "      <td>TTGCTGTGGGACTCAA</td>\n",
       "      <td>55.0</td>\n",
       "      <td>CAGCATTTGTTTCTGAA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.611300</td>\n",
       "      <td>...</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>943876</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>TAATCTAGCGCGACGTCTTTGAGTCCCACAGCAAAACTGCGACCGT...</td>\n",
       "      <td>TTCAGAAACAAATGCTGTTGAGTCCCACAGCAA</td>\n",
       "      <td>00033020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471968</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>1278.0</td>\n",
       "      <td>1278.0</td>\n",
       "      <td>ACCACTGCCCAACT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>CTATGCTTCTTGATACATG</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.552558</td>\n",
       "      <td>...</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>944102</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>TAATCTAGCGCGACGTCTAGTTGGGCAGTGGTAACTGCGACCGTCG...</td>\n",
       "      <td>CATGTATCAAGAAGCATAGAGTTGGGCAGTGGT</td>\n",
       "      <td>00033020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471969</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>3086.0</td>\n",
       "      <td>3086.0</td>\n",
       "      <td>TGGAGATGAATAAGTTGT</td>\n",
       "      <td>50.0</td>\n",
       "      <td>AAGGGTTGAGGCTAA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.485002</td>\n",
       "      <td>...</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>945910</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>TAATCTAGCGCGACGTCTACAACTTATTCATCTCCAAACTGCGACC...</td>\n",
       "      <td>TTAGCCTCAACCCTTACAACTTATTCATCTCCA</td>\n",
       "      <td>00033020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471970</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>3233.0</td>\n",
       "      <td>3233.0</td>\n",
       "      <td>TCCATGAGATAAGGCA</td>\n",
       "      <td>51.0</td>\n",
       "      <td>ACTGTAAGGCATTTTCT</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.493049</td>\n",
       "      <td>...</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>946057</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>TAATCTAGCGCGACGTCTTGCCTTATCTCATGGAAACTGCGACCGT...</td>\n",
       "      <td>AGAAAATGCCTTACAGTTGCCTTATCTCATGGA</td>\n",
       "      <td>00033020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471971</th>\n",
       "      <td>Kdm5d_chrY_26</td>\n",
       "      <td>chrY:000942824-946316</td>\n",
       "      <td>0</td>\n",
       "      <td>3338.0</td>\n",
       "      <td>3338.0</td>\n",
       "      <td>TCTATAAAAAGCAGGCTG</td>\n",
       "      <td>52.0</td>\n",
       "      <td>ATCAACCCAGGTGAG</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.528349</td>\n",
       "      <td>...</td>\n",
       "      <td>14s</td>\n",
       "      <td>Kdm5d</td>\n",
       "      <td>chrY</td>\n",
       "      <td>000942824</td>\n",
       "      <td>946316</td>\n",
       "      <td>946162</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>TAATCTAGCGCGACGTCTCAGCCTGCTTTTTATAGAAACTGCGACC...</td>\n",
       "      <td>CTCACCTGGGTTGATCAGCCTGCTTTTTATAGA</td>\n",
       "      <td>00033020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20195 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0                        1  2       3       4  \\\n",
       "7019      Il17a_chr1_1  chr1:020801129-20801212  0    65.0    65.0   \n",
       "7020      Il17a_chr1_2  chr1:020802320-20802531  0    69.0    69.0   \n",
       "7021      Il17a_chr1_2  chr1:020802320-20802531  0    93.0    93.0   \n",
       "7022      Il17a_chr1_2  chr1:020802320-20802531  0   112.0   112.0   \n",
       "7023      Il17a_chr1_2  chr1:020802320-20802531  0   130.0   130.0   \n",
       "...                ...                      ... ..     ...     ...   \n",
       "1471967  Kdm5d_chrY_26    chrY:000942824-946316  0  1052.0  1052.0   \n",
       "1471968  Kdm5d_chrY_26    chrY:000942824-946316  0  1278.0  1278.0   \n",
       "1471969  Kdm5d_chrY_26    chrY:000942824-946316  0  3086.0  3086.0   \n",
       "1471970  Kdm5d_chrY_26    chrY:000942824-946316  0  3233.0  3233.0   \n",
       "1471971  Kdm5d_chrY_26    chrY:000942824-946316  0  3338.0  3338.0   \n",
       "\n",
       "                          5     6                    7     8         9  ...  \\\n",
       "7019        GCAAACATGAGTCCA  50.0      GGGAGAGCTTCATCT  51.0  2.370986  ...   \n",
       "7020       ATCATCCCTCAAAGCT  51.0    CAGCGTGTCCAAACACT  58.0  2.416665  ...   \n",
       "7021         AAACACTGAGGCCA  50.0  AGGACTTCCTCCAGAATGT  58.0  2.437995  ...   \n",
       "7022      GACTTCCTCCAGAATGT  53.0     GAAGGTCAACCTCAAA  50.0  2.539507  ...   \n",
       "7023      AAGGTCAACCTCAAAGT  53.0     CTTTAACTCCCTTGGC  51.0  2.531687  ...   \n",
       "...                     ...   ...                  ...   ...       ...  ...   \n",
       "1471967    TTGCTGTGGGACTCAA  55.0    CAGCATTTGTTTCTGAA  50.0  2.611300  ...   \n",
       "1471968      ACCACTGCCCAACT  53.0  CTATGCTTCTTGATACATG  51.0  2.552558  ...   \n",
       "1471969  TGGAGATGAATAAGTTGT  50.0      AAGGGTTGAGGCTAA  50.0  2.485002  ...   \n",
       "1471970    TCCATGAGATAAGGCA  51.0    ACTGTAAGGCATTTTCT  50.0  2.493049  ...   \n",
       "1471971  TCTATAAAAAGCAGGCTG  52.0      ATCAACCCAGGTGAG  51.0  2.528349  ...   \n",
       "\n",
       "          13 gene_name   chr exon_start  exon_end probe_center Annotation  \\\n",
       "7019      1s     Il17a  chr1  020801129  20801212     20801194        CDS   \n",
       "7020      2s     Il17a  chr1  020802320  20802531     20802389        CDS   \n",
       "7021      2s     Il17a  chr1  020802320  20802531     20802413        CDS   \n",
       "7022      2s     Il17a  chr1  020802320  20802531     20802432        CDS   \n",
       "7023      2s     Il17a  chr1  020802320  20802531     20802450        CDS   \n",
       "...      ...       ...   ...        ...       ...          ...        ...   \n",
       "1471967  14s     Kdm5d  chrY  000942824    946316       943876      3'UTR   \n",
       "1471968  14s     Kdm5d  chrY  000942824    946316       944102      3'UTR   \n",
       "1471969  14s     Kdm5d  chrY  000942824    946316       945910      3'UTR   \n",
       "1471970  14s     Kdm5d  chrY  000942824    946316       946057      3'UTR   \n",
       "1471971  14s     Kdm5d  chrY  000942824    946316       946162      3'UTR   \n",
       "\n",
       "                                                full_probe  \\\n",
       "7019     TAATCTAGCGCGACGTCTTGGACTCATGTTTGCAGAACTTGCGCGG...   \n",
       "7020     TAATCTAGCGCGACGTCTAGCTTTGAGGGATGATAGAACTTGCGCG...   \n",
       "7021     TAATCTAGCGCGACGTCTTGGCCTCAGTGTTTAGAACTTGCGCGGA...   \n",
       "7022     TAATCTAGCGCGACGTCTACATTCTGGAGGAAGTCAGAACTTGCGC...   \n",
       "7023     TAATCTAGCGCGACGTCTACTTTGAGGTTGACCTTAGAACTTGCGC...   \n",
       "...                                                    ...   \n",
       "1471967  TAATCTAGCGCGACGTCTTTGAGTCCCACAGCAAAACTGCGACCGT...   \n",
       "1471968  TAATCTAGCGCGACGTCTAGTTGGGCAGTGGTAACTGCGACCGTCG...   \n",
       "1471969  TAATCTAGCGCGACGTCTACAACTTATTCATCTCCAAACTGCGACC...   \n",
       "1471970  TAATCTAGCGCGACGTCTTGCCTTATCTCATGGAAACTGCGACCGT...   \n",
       "1471971  TAATCTAGCGCGACGTCTCAGCCTGCTTTTTATAGAAACTGCGACC...   \n",
       "\n",
       "                          RNA_target_probe   barcode  \n",
       "7019        AGATGAAGCTCTCCCTGGACTCATGTTTGC  00102300  \n",
       "7020     AGTGTTTGGACACGCTGAGCTTTGAGGGATGAT  00102300  \n",
       "7021     ACATTCTGGAGGAAGTCCTTGGCCTCAGTGTTT  00102300  \n",
       "7022     TTTGAGGTTGACCTTCACATTCTGGAGGAAGTC  00102300  \n",
       "7023     GCCAAGGGAGTTAAAGACTTTGAGGTTGACCTT  00102300  \n",
       "...                                    ...       ...  \n",
       "1471967  TTCAGAAACAAATGCTGTTGAGTCCCACAGCAA  00033020  \n",
       "1471968  CATGTATCAAGAAGCATAGAGTTGGGCAGTGGT  00033020  \n",
       "1471969  TTAGCCTCAACCCTTACAACTTATTCATCTCCA  00033020  \n",
       "1471970  AGAAAATGCCTTACAGTTGCCTTATCTCATGGA  00033020  \n",
       "1471971  CTCACCTGGGTTGATCAGCCTGCTTTTTATAGA  00033020  \n",
       "\n",
       "[20195 rows x 23 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from itertools import permutations\n",
    "n_on = 3 # number of on-cycles!\n",
    "min_len = 152 # minimum length of synthesized probes\n",
    "non_resolveds = []\n",
    "\n",
    "# for i, gene in enumerate(allgenes):\n",
    "for i in barcodes_seq.index:\n",
    "    gene = barcodes_seq.loc[i, 'Gene name']\n",
    "    ppd_thisGene = all_ppd.loc[all_ppd['gene_name'] == gene]\n",
    "    \n",
    "    if ppd_thisGene.shape[0] == 0:\n",
    "        print(\"No probes for {}.\".format(gene))\n",
    "        continue\n",
    "    \n",
    "    for ind in ppd_thisGene.index: # for each probe of this gene\n",
    "        bare_len = (len(AP1V8) + len(anchorProbe) + len(AP2V8) + \n",
    "                    len(ppd_thisGene.loc[ind, '7']) + len(ppd_thisGene.loc[ind, '5']) +\n",
    "                    len(barcodes_seq.loc[i][0]) + len(barcodes_seq.loc[i][1]) + \n",
    "                    len(barcodes_seq.loc[i][2]))\n",
    "        n_Ts = min_len - bare_len\n",
    "        for j, perm in enumerate(permutations(range(n_on))): # check all permutations of barcodes\n",
    "            # fp = (AP1V8 +  # Conventional version\n",
    "            #       ppd_thisGene.loc[ind, '7'] +  \n",
    "            #       barcodes_seq.loc[i][perm[0]] + \n",
    "            #       ''.join(n_Ts * ['A']) + \n",
    "            #       anchorProbe + \n",
    "            #       barcodes_seq.loc[i][perm[1]] + \n",
    "            #       barcodes_seq.loc[i][perm[2]] +\n",
    "            #       ppd_thisGene.loc[ind, '5'] + \n",
    "            #       reverse_complement(AP2V8))\n",
    "\n",
    "            fp = (AP1V8 +  # SplintR version\n",
    "                  reverse_complement(ppd_thisGene.loc[ind, '5']) +  # changed to splintR version\n",
    "                  barcodes_seq.loc[i][perm[0]] + \n",
    "                  ''.join(n_Ts * ['A']) + \n",
    "                  anchorProbe + \n",
    "                  barcodes_seq.loc[i][perm[1]] + \n",
    "                  barcodes_seq.loc[i][perm[2]] +\n",
    "                  reverse_complement(ppd_thisGene.loc[ind, '7']) + # changed to splintR version\n",
    "                  reverse_complement(AP2V8))\n",
    "            RNA_target_probe = (reverse_complement(ppd_thisGene.loc[ind, '7']) +  reverse_complement(ppd_thisGene.loc[ind, '5']))\n",
    "            if not dpnii_site in fp[0:-len(AP2V8)]:\n",
    "#                 if j > 0:\n",
    "#                     print(\"new barcode rescued gene {0}, index {1}\".format(gene, ind))\n",
    "                break                \n",
    "\n",
    "            elif j == len(list(permutations(range(n_on)))) - 1:\n",
    "                non_resolveds.append(ind)\n",
    "                print(\"No DpnII free barcode for gene {0}, index {1}\".format(gene, ind))\n",
    "        all_ppd.loc[ind, 'full_probe'] = fp\n",
    "        all_ppd.loc[ind, 'RNA_target_probe'] = RNA_target_probe\n",
    "            \n",
    "    all_ppd.loc[all_ppd['gene_name'] == gene, 'barcode'] = barcodes_seq.loc[i]['barcode']     \n",
    "\n",
    "all_ppd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "2a3fa13a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Ccl27a', 'Cbx3', 'Bnip3', 'H3c6', 'Ccl27b'} genes have no good probes designed\n"
     ]
    }
   ],
   "source": [
    "print('{} genes have no good probes designed'.format(set(allgenes) - set(all_ppd['gene_name'])))\n",
    "allgenes = set(allgenes) & set(all_ppd['gene_name'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f993b5db-2987-4f67-b19d-2f49f082bec4",
   "metadata": {},
   "source": [
    "### select n_max probes for each gene according to the strategy above\n",
    "removing the probes that do have DpnII before the selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "8c9a5069",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67 probes are duplicated\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>13</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>probe_center</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>full_probe</th>\n",
       "      <th>RNA_target_probe</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>205745</th>\n",
       "      <td>Dnmt3b_chr2_3</td>\n",
       "      <td>chr2:153503345-153503522</td>\n",
       "      <td>0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>AACAATGAAGGGAGACA</td>\n",
       "      <td>52.0</td>\n",
       "      <td>GCAGACATCTGAATGA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.581573</td>\n",
       "      <td>...</td>\n",
       "      <td>1s</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153503345</td>\n",
       "      <td>153503522</td>\n",
       "      <td>153503363</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTTGTCTCCCTTCATTGTTTCGTACTTCGA...</td>\n",
       "      <td>TCATTCAGATGTCTGCTGTCTCCCTTCATTGTT</td>\n",
       "      <td>00020203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205746</th>\n",
       "      <td>Dnmt3b_chr2_3</td>\n",
       "      <td>chr2:153503345-153503522</td>\n",
       "      <td>0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>AGGAGTGCATTATCGTT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>AATGGGAACTTCAGTG</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.600552</td>\n",
       "      <td>...</td>\n",
       "      <td>1s</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153503345</td>\n",
       "      <td>153503522</td>\n",
       "      <td>153503419</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTAACGATAATGCACTCCTTCGTACTTCGA...</td>\n",
       "      <td>CACTGAAGTTCCCATTAACGATAATGCACTCCT</td>\n",
       "      <td>00020203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205749</th>\n",
       "      <td>Dnmt3b_chr2_3</td>\n",
       "      <td>chr2:153503345-153503522</td>\n",
       "      <td>0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>AGTCTTGGAGGCAAT</td>\n",
       "      <td>51.0</td>\n",
       "      <td>CTGCACAGAGCCAGTCT</td>\n",
       "      <td>59.0</td>\n",
       "      <td>2.263230</td>\n",
       "      <td>...</td>\n",
       "      <td>1s</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153503345</td>\n",
       "      <td>153503522</td>\n",
       "      <td>153503487</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTATTGCCTCCAAGACTTCGTACTTCGACG...</td>\n",
       "      <td>AGACTGGCTCTGTGCAGATTGCCTCCAAGACT</td>\n",
       "      <td>00020203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205751</th>\n",
       "      <td>Dnmt3b_chr2_4</td>\n",
       "      <td>chr2:153504063-153504124</td>\n",
       "      <td>0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>TCTAAGAGGGAGGTCT</td>\n",
       "      <td>51.0</td>\n",
       "      <td>CCAGCCTTCTGAATTA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.487027</td>\n",
       "      <td>...</td>\n",
       "      <td>0s</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153504063</td>\n",
       "      <td>153504124</td>\n",
       "      <td>153504101</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTAGACCTCCCTCTTAGAACGTCTGCGTAC...</td>\n",
       "      <td>TAATTCAGAAGGCTGGAGACCTCCCTCTTAGA</td>\n",
       "      <td>00020203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205752</th>\n",
       "      <td>Dnmt3b_chr2_5</td>\n",
       "      <td>chr2:153504634-153504753</td>\n",
       "      <td>0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>ACATGACAGGAGATGG</td>\n",
       "      <td>52.0</td>\n",
       "      <td>AGACAGAGATGATGAAG</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.532911</td>\n",
       "      <td>...</td>\n",
       "      <td>0s</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153504634</td>\n",
       "      <td>153504753</td>\n",
       "      <td>153504650</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTCCATCTCCTGTCATGTTCGTACTTCGAC...</td>\n",
       "      <td>CTTCATCATCTCTGTCTCCATCTCCTGTCATGT</td>\n",
       "      <td>00020203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479060</th>\n",
       "      <td>Col1a2_chr6_7</td>\n",
       "      <td>chr6:004515210-4515254</td>\n",
       "      <td>0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>TTTAATGGGACCCAGA</td>\n",
       "      <td>51.0</td>\n",
       "      <td>GGCCCTCCTGGTGC</td>\n",
       "      <td>59.0</td>\n",
       "      <td>1.832950</td>\n",
       "      <td>...</td>\n",
       "      <td>0s</td>\n",
       "      <td>Col1a2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>004515210</td>\n",
       "      <td>4515254</td>\n",
       "      <td>4515227</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTTCTGGGTCCCATTAAACGTATCGGTAGT...</td>\n",
       "      <td>GCACCAGGAGGGCCTCTGGGTCCCATTAAA</td>\n",
       "      <td>20000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479061</th>\n",
       "      <td>Col1a2_chr6_8</td>\n",
       "      <td>chr6:004515481-4515534</td>\n",
       "      <td>0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>TCAAGGTTTCCAAGGA</td>\n",
       "      <td>51.0</td>\n",
       "      <td>CCTGCTGGTGAACCT</td>\n",
       "      <td>55.0</td>\n",
       "      <td>2.220026</td>\n",
       "      <td>...</td>\n",
       "      <td>1s</td>\n",
       "      <td>Col1a2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>004515481</td>\n",
       "      <td>4515534</td>\n",
       "      <td>4515501</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTTCCTTGGAAACCTTGACGTATCGGTAGT...</td>\n",
       "      <td>AGGTTCACCAGCAGGTCCTTGGAAACCTTGA</td>\n",
       "      <td>20000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479062</th>\n",
       "      <td>Col1a2_chr6_9</td>\n",
       "      <td>chr6:004515629-4515682</td>\n",
       "      <td>0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>CAGCTGGCTCTCCT</td>\n",
       "      <td>54.0</td>\n",
       "      <td>GGCAAGGCTGGTGAG</td>\n",
       "      <td>57.0</td>\n",
       "      <td>1.380224</td>\n",
       "      <td>...</td>\n",
       "      <td>0s</td>\n",
       "      <td>Col1a2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>004515629</td>\n",
       "      <td>4515682</td>\n",
       "      <td>4515664</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTAGGAGAGCCAGCTGCGTATCGGTAGTCG...</td>\n",
       "      <td>CTCACCAGCCTTGCCAGGAGAGCCAGCTG</td>\n",
       "      <td>20000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479063</th>\n",
       "      <td>Col1a2_chr6_10</td>\n",
       "      <td>chr6:004515961-4516014</td>\n",
       "      <td>0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>TGGAAAACCCGGAA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>GACCTGGGGAGAGAGGA</td>\n",
       "      <td>59.0</td>\n",
       "      <td>2.100101</td>\n",
       "      <td>...</td>\n",
       "      <td>0s</td>\n",
       "      <td>Col1a2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>004515961</td>\n",
       "      <td>4516014</td>\n",
       "      <td>4515982</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTTTCCGGGTTTTCCACGTATCGGTAGTCG...</td>\n",
       "      <td>TCCTCTCTCCCCAGGTCTTCCGGGTTTTCCA</td>\n",
       "      <td>20000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479065</th>\n",
       "      <td>Col1a2_chr6_11</td>\n",
       "      <td>chr6:004516392-4516445</td>\n",
       "      <td>0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>ACTCCTGGACTTCCT</td>\n",
       "      <td>52.0</td>\n",
       "      <td>GGCTTCAAAGGCGTGAAA</td>\n",
       "      <td>59.0</td>\n",
       "      <td>2.303658</td>\n",
       "      <td>...</td>\n",
       "      <td>1s</td>\n",
       "      <td>Col1a2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>004516392</td>\n",
       "      <td>4516445</td>\n",
       "      <td>4516427</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTAGGAAGTCCAGGAGTCGTATCGGTAGTC...</td>\n",
       "      <td>TTTCACGCCTTTGAAGCCAGGAAGTCCAGGAGT</td>\n",
       "      <td>20000021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3060 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0                         1  2      3      4  \\\n",
       "205745   Dnmt3b_chr2_3  chr2:153503345-153503522  0   18.0   18.0   \n",
       "205746   Dnmt3b_chr2_3  chr2:153503345-153503522  0   74.0   74.0   \n",
       "205749   Dnmt3b_chr2_3  chr2:153503345-153503522  0  142.0  142.0   \n",
       "205751   Dnmt3b_chr2_4  chr2:153504063-153504124  0   38.0   38.0   \n",
       "205752   Dnmt3b_chr2_5  chr2:153504634-153504753  0   16.0   16.0   \n",
       "...                ...                       ... ..    ...    ...   \n",
       "479060   Col1a2_chr6_7    chr6:004515210-4515254  0   17.0   17.0   \n",
       "479061   Col1a2_chr6_8    chr6:004515481-4515534  0   20.0   20.0   \n",
       "479062   Col1a2_chr6_9    chr6:004515629-4515682  0   35.0   35.0   \n",
       "479063  Col1a2_chr6_10    chr6:004515961-4516014  0   21.0   21.0   \n",
       "479065  Col1a2_chr6_11    chr6:004516392-4516445  0   35.0   35.0   \n",
       "\n",
       "                        5     6                   7     8         9  ...  13  \\\n",
       "205745  AACAATGAAGGGAGACA  52.0    GCAGACATCTGAATGA  50.0  2.581573  ...  1s   \n",
       "205746  AGGAGTGCATTATCGTT  53.0    AATGGGAACTTCAGTG  50.0  2.600552  ...  1s   \n",
       "205749    AGTCTTGGAGGCAAT  51.0   CTGCACAGAGCCAGTCT  59.0  2.263230  ...  1s   \n",
       "205751   TCTAAGAGGGAGGTCT  51.0    CCAGCCTTCTGAATTA  50.0  2.487027  ...  0s   \n",
       "205752   ACATGACAGGAGATGG  52.0   AGACAGAGATGATGAAG  50.0  2.532911  ...  0s   \n",
       "...                   ...   ...                 ...   ...       ...  ...  ..   \n",
       "479060   TTTAATGGGACCCAGA  51.0      GGCCCTCCTGGTGC  59.0  1.832950  ...  0s   \n",
       "479061   TCAAGGTTTCCAAGGA  51.0     CCTGCTGGTGAACCT  55.0  2.220026  ...  1s   \n",
       "479062     CAGCTGGCTCTCCT  54.0     GGCAAGGCTGGTGAG  57.0  1.380224  ...  0s   \n",
       "479063     TGGAAAACCCGGAA  50.0   GACCTGGGGAGAGAGGA  59.0  2.100101  ...  0s   \n",
       "479065    ACTCCTGGACTTCCT  52.0  GGCTTCAAAGGCGTGAAA  59.0  2.303658  ...  1s   \n",
       "\n",
       "       gene_name   chr exon_start   exon_end probe_center Annotation  \\\n",
       "205745    Dnmt3b  chr2  153503345  153503522    153503363        CDS   \n",
       "205746    Dnmt3b  chr2  153503345  153503522    153503419        CDS   \n",
       "205749    Dnmt3b  chr2  153503345  153503522    153503487        CDS   \n",
       "205751    Dnmt3b  chr2  153504063  153504124    153504101        CDS   \n",
       "205752    Dnmt3b  chr2  153504634  153504753    153504650        CDS   \n",
       "...          ...   ...        ...        ...          ...        ...   \n",
       "479060    Col1a2  chr6  004515210    4515254      4515227        CDS   \n",
       "479061    Col1a2  chr6  004515481    4515534      4515501        CDS   \n",
       "479062    Col1a2  chr6  004515629    4515682      4515664        CDS   \n",
       "479063    Col1a2  chr6  004515961    4516014      4515982        CDS   \n",
       "479065    Col1a2  chr6  004516392    4516445      4516427        CDS   \n",
       "\n",
       "                                               full_probe  \\\n",
       "205745  TAATCTAGCGCGACGTCTTGTCTCCCTTCATTGTTTCGTACTTCGA...   \n",
       "205746  TAATCTAGCGCGACGTCTAACGATAATGCACTCCTTCGTACTTCGA...   \n",
       "205749  TAATCTAGCGCGACGTCTATTGCCTCCAAGACTTCGTACTTCGACG...   \n",
       "205751  TAATCTAGCGCGACGTCTAGACCTCCCTCTTAGAACGTCTGCGTAC...   \n",
       "205752  TAATCTAGCGCGACGTCTCCATCTCCTGTCATGTTCGTACTTCGAC...   \n",
       "...                                                   ...   \n",
       "479060  TAATCTAGCGCGACGTCTTCTGGGTCCCATTAAACGTATCGGTAGT...   \n",
       "479061  TAATCTAGCGCGACGTCTTCCTTGGAAACCTTGACGTATCGGTAGT...   \n",
       "479062  TAATCTAGCGCGACGTCTAGGAGAGCCAGCTGCGTATCGGTAGTCG...   \n",
       "479063  TAATCTAGCGCGACGTCTTTCCGGGTTTTCCACGTATCGGTAGTCG...   \n",
       "479065  TAATCTAGCGCGACGTCTAGGAAGTCCAGGAGTCGTATCGGTAGTC...   \n",
       "\n",
       "                         RNA_target_probe   barcode  \n",
       "205745  TCATTCAGATGTCTGCTGTCTCCCTTCATTGTT  00020203  \n",
       "205746  CACTGAAGTTCCCATTAACGATAATGCACTCCT  00020203  \n",
       "205749   AGACTGGCTCTGTGCAGATTGCCTCCAAGACT  00020203  \n",
       "205751   TAATTCAGAAGGCTGGAGACCTCCCTCTTAGA  00020203  \n",
       "205752  CTTCATCATCTCTGTCTCCATCTCCTGTCATGT  00020203  \n",
       "...                                   ...       ...  \n",
       "479060     GCACCAGGAGGGCCTCTGGGTCCCATTAAA  20000021  \n",
       "479061    AGGTTCACCAGCAGGTCCTTGGAAACCTTGA  20000021  \n",
       "479062      CTCACCAGCCTTGCCAGGAGAGCCAGCTG  20000021  \n",
       "479063    TCCTCTCTCCCCAGGTCTTCCGGGTTTTCCA  20000021  \n",
       "479065  TTTCACGCCTTTGAAGCCAGGAAGTCCAGGAGT  20000021  \n",
       "\n",
       "[3060 rows x 23 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select probes based on the order: CDS, 5'UTR, 3'UTR and prioritize probes close to TSS and non-overlapping probes\n",
    "\n",
    "dist_for_nonoverlap = 33\n",
    "selprobes = []\n",
    "for g in set(allgenes):\n",
    "# for g in ['Smarca5','Il17a']:\n",
    "    ppd_g = all_ppd.loc[~all_ppd.index.isin(non_resolveds)].query(\"gene_name == @g\")\n",
    "    ppd_g = ppd_g.sort_values('probe_center')\n",
    "    n_probe = ppd_g.shape[0]\n",
    "    if n_probe <= n_max:\n",
    "        sel_index = ppd_g.index\n",
    "        sel_index = sel_index.append(ppd_g.sample(n = n_max - len(sel_index), replace=True).index)\n",
    "        sel_g = ppd_g.loc[sel_index]\n",
    "    else:\n",
    "        sel_index = []\n",
    "        sel_probe_center = []\n",
    "        while len(sel_index) < n_max:\n",
    "            if ppd_g['10'].iloc[0] == 'W': # Walson strand or + strand, select probes close to TSS - smaller probe_center\n",
    "                for annot in ['CDS', \"5'UTR\", \"3'UTR\"]:\n",
    "                    if len(ppd_g.loc[ppd_g['Annotation'] == annot]) != 0 and len(sel_index) == 0:\n",
    "                        sel_index.append(ppd_g.loc[ppd_g['Annotation'] == annot].index[0])\n",
    "                        sel_probe_center.append(ppd_g.loc[ppd_g['Annotation'] == annot]['probe_center'].iloc[0])\n",
    "                        ppd_g = ppd_g.drop(ppd_g.loc[ppd_g['Annotation'] == annot].index[0])\n",
    "                    if len(sel_index) < n_max:\n",
    "                        for index in ppd_g.loc[ppd_g['Annotation'] == annot].index:\n",
    "                            probe_center = ppd_g.loc[index, 'probe_center']\n",
    "                            if all(np.abs(probe_center - np.array(sel_probe_center)) > dist_for_nonoverlap):\n",
    "                                sel_index.append(index)\n",
    "                                sel_probe_center.append(probe_center)\n",
    "                                ppd_g = ppd_g.drop(index)\n",
    "                            if len(sel_index) == n_max:\n",
    "                                break\n",
    "                if len(sel_index) < n_max:\n",
    "                    sel_index.extend(ppd_g.sample(n = n_max - len(sel_index), replace=True).index.values)\n",
    "            else: # Crick strand or - strand, select probes close to TSS - larger probe_center\n",
    "                for annot in ['CDS', \"5'UTR\", \"3'UTR\"]:\n",
    "                    if len(ppd_g.loc[ppd_g['Annotation'] == annot]) != 0 and len(sel_index) == 0:\n",
    "                        sel_index.append(ppd_g.loc[ppd_g['Annotation'] == annot].index[-1])\n",
    "                        sel_probe_center.append(ppd_g.loc[ppd_g['Annotation'] == annot]['probe_center'].iloc[-1])\n",
    "                        ppd_g = ppd_g.drop(ppd_g.loc[ppd_g['Annotation'] == annot].index[-1])\n",
    "                    if len(sel_index) < n_max:\n",
    "                        for index in ppd_g.loc[ppd_g['Annotation'] == annot].index[::-1]:\n",
    "                            probe_center = ppd_g.loc[index, 'probe_center']\n",
    "                            if all(np.abs(probe_center - np.array(sel_probe_center)) > dist_for_nonoverlap):\n",
    "                                sel_index.append(index)\n",
    "                                sel_probe_center.append(probe_center)\n",
    "                                ppd_g = ppd_g.drop(index)\n",
    "                            if len(sel_index) == n_max:\n",
    "                                break\n",
    "                if len(sel_index) < n_max:\n",
    "                    sel_index.extend(ppd_g.sample(n = n_max - len(sel_index), replace=True).index.values)\n",
    "        ppd_g = all_ppd.loc[~all_ppd.index.isin(non_resolveds)].query(\"gene_name == @g\")\n",
    "        sel_g = ppd_g.loc[sel_index]\n",
    "    selprobes.append(sel_g)\n",
    "selprobes = pd.concat(selprobes)\n",
    "print(\"{} probes are duplicated\".format(selprobes.shape[0] - selprobes.index.unique().shape[0]))\n",
    "selprobes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "6f203402-5aa6-4b6f-95cb-524c70ef3ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# selprobes = []\n",
    "# for g in set(allgenes) - set(['Ccl27a','Cbx3','Bnip3']):\n",
    "#     ppd_g = all_ppd.loc[~all_ppd.index.isin(non_resolveds)].query(\"gene_name == @g\")\n",
    "#     ppd_g_GGGG_free = ppd_g.loc[(~ppd_g['7'].str.contains('CCCC')) & (~ppd_g['5'].str.contains('CCCC'))]\n",
    "#     n_l = (ppd_g_GGGG_free['subset'] == 'l').sum()\n",
    "#     n_s = (ppd_g_GGGG_free['subset'] == 's').sum()\n",
    "#     if n_l > n_max: # too many non-overlapping probes\n",
    "#         sel_g = ppd_g_GGGG_free.query(\"subset == 'l'\").sample(n = n_max, replace=False)\n",
    "#     elif n_l + n_s > n_max: # select all of l-subset and fill up with s-subset\n",
    "#         s_index = ppd_g_GGGG_free.query(\"subset == 's'\").sample(n = n_max-n_l, replace=False).index\n",
    "#         sel_index = ppd_g_GGGG_free.query(\"subset=='l'\").index.append(s_index)\n",
    "#         sel_g = ppd_g_GGGG_free.loc[sel_index]\n",
    "#     else: # less than n_max probes \n",
    "#         sel_index = ppd_g_GGGG_free.index\n",
    "#         sel_index = sel_index.append(ppd_g_GGGG_free.sample(n = n_max - len(sel_index), replace=True).index)\n",
    "#         sel_g = ppd_g_GGGG_free.loc[sel_index]\n",
    "#     selprobes.append(sel_g)\n",
    "# selprobes = pd.concat(selprobes)\n",
    "# print(\"{} probes are duplicated\".format(selprobes.shape[0] - selprobes.index.unique().shape[0]))\n",
    "# selprobes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "35ef84a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Annotation\n",
       "CDS      2734\n",
       "3'UTR     202\n",
       "5'UTR     124\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selprobes['Annotation'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "683d40dd-2673-4bce-a14a-2ffd7138696f",
   "metadata": {},
   "source": [
    "#### writing the probes to file in .ppd format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1ad58509-6b36-4b0b-b284-f97a20865a11",
   "metadata": {},
   "outputs": [],
   "source": [
    "unResolved = all_ppd['full_probe'].str.slice(0, -len(AP2V8)).str.contains(dpnii_site)\n",
    "unResolved = all_ppd.loc[unResolved]\n",
    "unResolved[['gene_name', 'barcode', 'full_probe']].to_csv('2408_MCrossTissue_probesWithGATC_3col.tsv', sep = '\\t', index = False)\n",
    "all_ppd = all_ppd.drop(unResolved.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "f00698bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Hs6st3_chr14_1', 'chr14:119375753-119376530', 0, 78.0, 78.0,\n",
       "       'AACGGGACCATGGA', 53.0, 'TGAAAGGTTCAACAAGT', 51.0, 2.542873, 'W',\n",
       "       'AACGGGACCATGGATGAAAGGTTCAACAAGT', 0.1, '0s', 'Hs6st3', 'chr14',\n",
       "       '119375753', '119376530', 119375831, 'CDS',\n",
       "       'TAATCTAGCGCGACGTCTTCCATGGTCCCGTTCGTATCGGTAGTCGCAACGCAAACTTCAGCTTCCCGATATCCGGACGAACGGTCGAGATTTACAACTGCGACCGTCGGCTTACACTTGTTGAACCTTTCAGATCATAGCGCCTCTTGTGG',\n",
       "       'ACTTGTTGAACCTTTCATCCATGGTCCCGTT', '20330000'], dtype=object)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selprobes.values[10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e086d21-a3b7-42bd-919e-39f61624709c",
   "metadata": {},
   "source": [
    "### add \"A\" to the 5' to get them to ```target_len``` length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3965e54e-b845-4375-84bd-44d0540564ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "205745    TAATCTAGCGCGACGTCTTGTCTCCCTTCATTGTTTCGTACTTCGA...\n",
       "205746    TAATCTAGCGCGACGTCTAACGATAATGCACTCCTTCGTACTTCGA...\n",
       "205749    TAATCTAGCGCGACGTCTATTGCCTCCAAGACTTCGTACTTCGACG...\n",
       "205751    TAATCTAGCGCGACGTCTAGACCTCCCTCTTAGAACGTCTGCGTAC...\n",
       "205752    TAATCTAGCGCGACGTCTCCATCTCCTGTCATGTTCGTACTTCGAC...\n",
       "Name: full_probe, dtype: object"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selprobes['full_probe'] = selprobes['full_probe'].str.pad(target_len, side='left', fillchar='A')\n",
    "selprobes['full_probe'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "185d02da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>exon_start</th>\n",
       "      <th>exon_end</th>\n",
       "      <th>probe_center</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>full_probe</th>\n",
       "      <th>RNA_target_probe</th>\n",
       "      <th>barcode</th>\n",
       "      <th>fasta_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>205745</th>\n",
       "      <td>Dnmt3b_chr2_3</td>\n",
       "      <td>chr2:153503345-153503522</td>\n",
       "      <td>0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>AACAATGAAGGGAGACA</td>\n",
       "      <td>52.0</td>\n",
       "      <td>GCAGACATCTGAATGA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.581573</td>\n",
       "      <td>...</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153503345</td>\n",
       "      <td>153503522</td>\n",
       "      <td>153503363</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTTGTCTCCCTTCATTGTTTCGTACTTCGA...</td>\n",
       "      <td>TCATTCAGATGTCTGCTGTCTCCCTTCATTGTT</td>\n",
       "      <td>00020203</td>\n",
       "      <td>Dnmt3b_00020203_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205746</th>\n",
       "      <td>Dnmt3b_chr2_3</td>\n",
       "      <td>chr2:153503345-153503522</td>\n",
       "      <td>0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>AGGAGTGCATTATCGTT</td>\n",
       "      <td>53.0</td>\n",
       "      <td>AATGGGAACTTCAGTG</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.600552</td>\n",
       "      <td>...</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153503345</td>\n",
       "      <td>153503522</td>\n",
       "      <td>153503419</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTAACGATAATGCACTCCTTCGTACTTCGA...</td>\n",
       "      <td>CACTGAAGTTCCCATTAACGATAATGCACTCCT</td>\n",
       "      <td>00020203</td>\n",
       "      <td>Dnmt3b_00020203_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205749</th>\n",
       "      <td>Dnmt3b_chr2_3</td>\n",
       "      <td>chr2:153503345-153503522</td>\n",
       "      <td>0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>AGTCTTGGAGGCAAT</td>\n",
       "      <td>51.0</td>\n",
       "      <td>CTGCACAGAGCCAGTCT</td>\n",
       "      <td>59.0</td>\n",
       "      <td>2.263230</td>\n",
       "      <td>...</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153503345</td>\n",
       "      <td>153503522</td>\n",
       "      <td>153503487</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTATTGCCTCCAAGACTTCGTACTTCGACG...</td>\n",
       "      <td>AGACTGGCTCTGTGCAGATTGCCTCCAAGACT</td>\n",
       "      <td>00020203</td>\n",
       "      <td>Dnmt3b_00020203_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205751</th>\n",
       "      <td>Dnmt3b_chr2_4</td>\n",
       "      <td>chr2:153504063-153504124</td>\n",
       "      <td>0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>TCTAAGAGGGAGGTCT</td>\n",
       "      <td>51.0</td>\n",
       "      <td>CCAGCCTTCTGAATTA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.487027</td>\n",
       "      <td>...</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153504063</td>\n",
       "      <td>153504124</td>\n",
       "      <td>153504101</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTAGACCTCCCTCTTAGAACGTCTGCGTAC...</td>\n",
       "      <td>TAATTCAGAAGGCTGGAGACCTCCCTCTTAGA</td>\n",
       "      <td>00020203</td>\n",
       "      <td>Dnmt3b_00020203_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205752</th>\n",
       "      <td>Dnmt3b_chr2_5</td>\n",
       "      <td>chr2:153504634-153504753</td>\n",
       "      <td>0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>ACATGACAGGAGATGG</td>\n",
       "      <td>52.0</td>\n",
       "      <td>AGACAGAGATGATGAAG</td>\n",
       "      <td>50.0</td>\n",
       "      <td>2.532911</td>\n",
       "      <td>...</td>\n",
       "      <td>Dnmt3b</td>\n",
       "      <td>chr2</td>\n",
       "      <td>153504634</td>\n",
       "      <td>153504753</td>\n",
       "      <td>153504650</td>\n",
       "      <td>CDS</td>\n",
       "      <td>TAATCTAGCGCGACGTCTCCATCTCCTGTCATGTTCGTACTTCGAC...</td>\n",
       "      <td>CTTCATCATCTCTGTCTCCATCTCCTGTCATGT</td>\n",
       "      <td>00020203</td>\n",
       "      <td>Dnmt3b_00020203_5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0                         1  2      3      4  \\\n",
       "205745  Dnmt3b_chr2_3  chr2:153503345-153503522  0   18.0   18.0   \n",
       "205746  Dnmt3b_chr2_3  chr2:153503345-153503522  0   74.0   74.0   \n",
       "205749  Dnmt3b_chr2_3  chr2:153503345-153503522  0  142.0  142.0   \n",
       "205751  Dnmt3b_chr2_4  chr2:153504063-153504124  0   38.0   38.0   \n",
       "205752  Dnmt3b_chr2_5  chr2:153504634-153504753  0   16.0   16.0   \n",
       "\n",
       "                        5     6                  7     8         9  ...  \\\n",
       "205745  AACAATGAAGGGAGACA  52.0   GCAGACATCTGAATGA  50.0  2.581573  ...   \n",
       "205746  AGGAGTGCATTATCGTT  53.0   AATGGGAACTTCAGTG  50.0  2.600552  ...   \n",
       "205749    AGTCTTGGAGGCAAT  51.0  CTGCACAGAGCCAGTCT  59.0  2.263230  ...   \n",
       "205751   TCTAAGAGGGAGGTCT  51.0   CCAGCCTTCTGAATTA  50.0  2.487027  ...   \n",
       "205752   ACATGACAGGAGATGG  52.0  AGACAGAGATGATGAAG  50.0  2.532911  ...   \n",
       "\n",
       "       gene_name   chr  exon_start   exon_end probe_center Annotation  \\\n",
       "205745    Dnmt3b  chr2   153503345  153503522    153503363        CDS   \n",
       "205746    Dnmt3b  chr2   153503345  153503522    153503419        CDS   \n",
       "205749    Dnmt3b  chr2   153503345  153503522    153503487        CDS   \n",
       "205751    Dnmt3b  chr2   153504063  153504124    153504101        CDS   \n",
       "205752    Dnmt3b  chr2   153504634  153504753    153504650        CDS   \n",
       "\n",
       "                                               full_probe  \\\n",
       "205745  TAATCTAGCGCGACGTCTTGTCTCCCTTCATTGTTTCGTACTTCGA...   \n",
       "205746  TAATCTAGCGCGACGTCTAACGATAATGCACTCCTTCGTACTTCGA...   \n",
       "205749  TAATCTAGCGCGACGTCTATTGCCTCCAAGACTTCGTACTTCGACG...   \n",
       "205751  TAATCTAGCGCGACGTCTAGACCTCCCTCTTAGAACGTCTGCGTAC...   \n",
       "205752  TAATCTAGCGCGACGTCTCCATCTCCTGTCATGTTCGTACTTCGAC...   \n",
       "\n",
       "                         RNA_target_probe   barcode         fasta_name  \n",
       "205745  TCATTCAGATGTCTGCTGTCTCCCTTCATTGTT  00020203  Dnmt3b_00020203_1  \n",
       "205746  CACTGAAGTTCCCATTAACGATAATGCACTCCT  00020203  Dnmt3b_00020203_2  \n",
       "205749   AGACTGGCTCTGTGCAGATTGCCTCCAAGACT  00020203  Dnmt3b_00020203_3  \n",
       "205751   TAATTCAGAAGGCTGGAGACCTCCCTCTTAGA  00020203  Dnmt3b_00020203_4  \n",
       "205752  CTTCATCATCTCTGTCTCCATCTCCTGTCATGT  00020203  Dnmt3b_00020203_5  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# forming the fasta name : gene_name + barcode + probe#\n",
    "selprobes['fasta_name'] = (selprobes['gene_name'] + '_' + \n",
    "                            selprobes['barcode'] + '_' +\n",
    "                            (selprobes.groupby('barcode').cumcount() + 1).astype(str))\n",
    "# writing the fasta name:\n",
    "with open(\"2408_MCrossTissue_probes_3col_V8.fa\", 'w') as writer:\n",
    "    for i, row in selprobes.iterrows():\n",
    "        writer.write('>{0}\\n'.format(row['fasta_name'])) # write the name\n",
    "        writer.write(\"{0}\\n\".format(row['full_probe']))\n",
    "\n",
    "selprobes[['fasta_name', 'full_probe']].to_csv(\"2408_MCrossTissue_probes_3col_V8_summary.tsv\", sep = '\\t', index = False)\n",
    "selprobes.to_csv(\"2408_MCrossTissue_probes_3col_V8_full.ppd\", sep = '\\t', index = False)\n",
    "selprobes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "331a43b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "codebook = (selprobes['gene_name'] + '_' + selprobes['barcode']).unique()\n",
    "with open(\"./2408_MCrossTissue_codebook_3col.txt\", \"w\") as writer:\n",
    "    for c in codebook:\n",
    "        writer.writelines(\"{}\\n\".format(c))\n",
    "Empty_codebook = empty_df['Gene name'] + '_' + empty_df.iloc[:,1:8].astype(str).agg(''.join, axis=1)\n",
    "with open(\"./2408_MCrossTissue_codebook_3col.txt\", \"a\") as writer:\n",
    "    for c in Empty_codebook:\n",
    "        writer.writelines(\"{}\\n\".format(c))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74b41a8a",
   "metadata": {},
   "source": [
    "### Making sure that the combination of these barcodes is recoverable "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "bb85b1f0",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: './short_genes_codebook.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[38], line 18\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m np\u001b[38;5;241m.\u001b[39marray(oh)\n\u001b[1;32m     17\u001b[0m cb_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m./short_genes_codebook.txt\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m---> 18\u001b[0m codebook \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcb_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m_\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnames\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgene\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbarcode\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     19\u001b[0m codebook[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124monehot\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m codebook[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbarcode\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mapply(onehot)\n\u001b[1;32m     20\u001b[0m display(codebook)\n",
      "File \u001b[0;32m/usr/conda/envs/py3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m    899\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    900\u001b[0m     dialect,\n\u001b[1;32m    901\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    908\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m    909\u001b[0m )\n\u001b[1;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/conda/envs/py3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    574\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m/usr/conda/envs/py3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1404\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/conda/envs/py3/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1661\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1659\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1660\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1662\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1663\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1664\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1665\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1668\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1670\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1671\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1672\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m/usr/conda/envs/py3/lib/python3.11/site-packages/pandas/io/common.py:859\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    854\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    855\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    856\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    857\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    858\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 859\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(\n\u001b[1;32m    860\u001b[0m             handle,\n\u001b[1;32m    861\u001b[0m             ioargs\u001b[38;5;241m.\u001b[39mmode,\n\u001b[1;32m    862\u001b[0m             encoding\u001b[38;5;241m=\u001b[39mioargs\u001b[38;5;241m.\u001b[39mencoding,\n\u001b[1;32m    863\u001b[0m             errors\u001b[38;5;241m=\u001b[39merrors,\n\u001b[1;32m    864\u001b[0m             newline\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    865\u001b[0m         )\n\u001b[1;32m    866\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    867\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    868\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './short_genes_codebook.txt'"
     ]
    }
   ],
   "source": [
    "import os, pandas as pd, numpy as np\n",
    "\n",
    "def onehot(bc):\n",
    "    bc = list(bc)\n",
    "    oh = []\n",
    "    for char in bc:\n",
    "        if char == '1':\n",
    "            oh += [1, 0, 0]\n",
    "        elif char == '2': \n",
    "            oh += [0, 1, 0]\n",
    "        elif char == '3':\n",
    "            oh += [0, 0, 1]\n",
    "        else:\n",
    "            oh += [0, 0, 0]\n",
    "    return np.array(oh)\n",
    "\n",
    "cb_file = \"./short_genes_codebook.txt\"\n",
    "codebook = pd.read_csv(cb_file, sep=\"_\", header=None, names=['gene', 'barcode'], dtype=str)\n",
    "codebook['onehot'] = codebook['barcode'].apply(onehot)\n",
    "display(codebook)\n",
    "\n",
    "from itertools import combinations\n",
    "\n",
    "mod2_dict = {}\n",
    "nomod_dict = {}\n",
    "for i, (g1, g2) in enumerate(combinations(codebook['gene'], 2)):\n",
    "    bc1 = codebook.query('gene==@g1')['onehot'].iloc[0]\n",
    "    bc2 = codebook.query('gene==@g2')['onehot'].iloc[0]\n",
    "    nomod = bc1 + bc2\n",
    "    mod2 = (nomod>0).astype(int)\n",
    "    nomod_key = ''.join(nomod.astype(str))\n",
    "    mod2_key = ''.join(mod2.astype(str))\n",
    "    if mod2_key in mod2_dict:\n",
    "        mod2_dict[mod2_key] += [(g1, g2)]\n",
    "    else:\n",
    "        mod2_dict[mod2_key] = [(g1, g2)]\n",
    "    if nomod_key in nomod_dict:\n",
    "        nomod_dict[nomod_key] += [(g1, g2)]\n",
    "    else:\n",
    "        nomod_dict[nomod_key] = [(g1, g2)]\n",
    "    \n",
    "    if i%1000 == 0:\n",
    "        print(i)\n",
    "        \n",
    "mod2_counts = {key: len(val) for key, val in mod2_dict.items()}\n",
    "mod2_df = pd.DataFrame.from_dict(mod2_counts, orient='index', columns=['count'])\n",
    "mod2_df['gene_pairs'] = [val for key, val in mod2_dict.items()]\n",
    "mod2_df = mod2_df.sort_values(by='count', ascending=False)\n",
    "display(mod2_df)\n",
    "\n",
    "nomod_counts = {key: len(val) for key, val in nomod_dict.items()}\n",
    "nomod_df = pd.DataFrame.from_dict(nomod_counts, orient='index', columns=['count'])\n",
    "nomod_df['gene_pairs'] = [val for key, val in nomod_dict.items()]\n",
    "nomod_df = nomod_df.sort_values(by='count', ascending=False)\n",
    "display(nomod_df)\n",
    "\n",
    "\n",
    "print(len(list(combinations(codebook['gene'], 2))))\n",
    "display(mod2_df['count'].value_counts())\n",
    "\n",
    "display(nomod_df['count'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e55c0e40",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc1 = codebook.query('gene==\"LINC02218\"')['onehot'].iloc[0]\n",
    "bc2 = codebook.query('gene==\"CMTM8\"')['onehot'].iloc[0]\n",
    "nomod = bc1 + bc2\n",
    "mod2 = (nomod>0).astype(int)\n",
    "nomod_key = ''.join(nomod.astype(str))\n",
    "mod2_key = ''.join(mod2.astype(str))\n",
    "# mod2_df[mod2_key]\n",
    "print(mod2_key)\n",
    "display(mod2_df.loc[mod2_key])\n",
    "print(nomod_key)\n",
    "display(nomod_df.loc[nomod_key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2a533d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, pairlist in enumerate(mod2_df['gene_pairs']):\n",
    "    for pair in pairlist:\n",
    "        if 'CMTM8' in pair:\n",
    "            n = mod2_df.iloc[i, 0]\n",
    "            if n > 1:\n",
    "                print(pairlist)\n",
    "                break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72cba649",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, pairlist in enumerate(mod2_df['gene_pairs']):\n",
    "    for pair in pairlist:\n",
    "        if 'TESC' in pair:\n",
    "            n = mod2_df.iloc[i, 0]\n",
    "            if n > 1:\n",
    "                print(pairlist)\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1c6b59a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os, pandas as pd, numpy as np\n",
    "\n",
    "def onehot(bc):\n",
    "    bc = list(bc)\n",
    "    oh = []\n",
    "    for char in bc:\n",
    "        if char == '1':\n",
    "            oh += [1, 0, 0]\n",
    "        elif char == '2': \n",
    "            oh += [0, 1, 0]\n",
    "        elif char == '3':\n",
    "            oh += [0, 0, 1]\n",
    "        else:\n",
    "            oh += [0, 0, 0]\n",
    "    return np.array(oh)\n",
    "\n",
    "cb_file = \"./short_genes_codebook.txt\"\n",
    "codebook = pd.read_csv(cb_file, sep=\"_\", header=None, names=['gene', 'barcode'], dtype=str)\n",
    "codebook['onehot'] = codebook['barcode'].apply(onehot)\n",
    "display(codebook)\n",
    "\n",
    "from itertools import combinations\n",
    "\n",
    "mod2_dict = {}\n",
    "nomod_dict = {}\n",
    "for i, (g1, g2, g3) in enumerate(combinations(codebook['gene'], 3)):\n",
    "    bc1 = codebook.query('gene==@g1')['onehot'].iloc[0]\n",
    "    bc2 = codebook.query('gene==@g2')['onehot'].iloc[0]\n",
    "    bc3 = codebook.query('gene==@g3')['onehot'].iloc[0]    \n",
    "    nomod = bc1 + bc2 + bc3\n",
    "    mod2 = (nomod>0).astype(int)\n",
    "    nomod_key = ''.join(nomod.astype(str))\n",
    "    mod2_key = ''.join(mod2.astype(str))\n",
    "    if mod2_key in mod2_dict:\n",
    "        mod2_dict[mod2_key] += [(g1, g2, g3)]\n",
    "    else:\n",
    "        mod2_dict[mod2_key] = [(g1, g2, g3)]\n",
    "    if nomod_key in nomod_dict:\n",
    "        nomod_dict[nomod_key] += [(g1, g2, g3)]\n",
    "    else:\n",
    "        nomod_dict[nomod_key] = [(g1, g2, g3)]\n",
    "    \n",
    "    if i%1000 == 0:\n",
    "        print(i)\n",
    "        \n",
    "mod2_counts = {key: len(val) for key, val in mod2_dict.items()}\n",
    "mod2_df = pd.DataFrame.from_dict(mod2_counts, orient='index', columns=['count'])\n",
    "mod2_df['gene_pairs'] = [val for key, val in mod2_dict.items()]\n",
    "mod2_df = mod2_df.sort_values(by='count', ascending=False)\n",
    "display(mod2_df)\n",
    "\n",
    "nomod_counts = {key: len(val) for key, val in nomod_dict.items()}\n",
    "nomod_df = pd.DataFrame.from_dict(nomod_counts, orient='index', columns=['count'])\n",
    "nomod_df['gene_pairs'] = [val for key, val in nomod_dict.items()]\n",
    "nomod_df = nomod_df.sort_values(by='count', ascending=False)\n",
    "display(nomod_df)\n",
    "\n",
    "\n",
    "print(len(list(combinations(codebook['gene'], 3))))\n",
    "display(mod2_df['count'].value_counts())\n",
    "\n",
    "display(nomod_df['count'].value_counts())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "image",
   "language": "python",
   "name": "image"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
